File: 1 - Algorithmic Toolbox/Week_3/2 - Maximum Value of the Loot/max_value_of_loot.py

def max_value( values: list[float], weights: list[float], max_capacity: int ) -> float:
    value = 0.0

    cost = [ v/w for v, w in zip( values, weights ) ]
    sort_lists = [ i for _, i in sorted( zip(cost, range( len( cost ) ) ), reverse=True)]

    for i in range( len( cost ) ):
        sorted_index = sort_lists[i]

        if max_capacity == 0:
            return value
        
        capacity_intake = min( weights[sorted_index], max_capacity )
        max_capacity -= capacity_intake
        value += capacity_intake * cost[sorted_index]
    
    return value

def main( ) -> None:
    num_items, max_capacity = map( int, input( ).split( ) )
    values = []
    weights = []
    for i in range( num_items ):
        user_input = input( )
        value, weight = map( int, user_input.split( " " ) )
        values.append( value )
        weights.append( weight )
    
    print( f'{max_value( values, weights, max_capacity ):.4f}' )

if __name__ == "__main__":
    main( )

File: 1 - Algorithmic Toolbox/3 - Greedy Algorithms/2 - Maximum Value of the Loot/max_value_of_loot.c
#include <stdio.h>
#include <stdlib.h>

double max_value(double values[], double weights[], int num_items, int max_capacity) 
{
    double value = 0.0 ;

    // Calculate the cost (value/weight ratio) for each item
    double *cost = (double *) malloc(num_items * sizeof(double)) ;
    int *sort_indexes = (int *) malloc(num_items * sizeof(int)) ;

    for (int i = 0; i < num_items; i++) {
        cost[i] = values[i] / weights[i] ;
        sort_indexes[i] = i ;
    }

    for (int i = 0; i < num_items - 1; i++) {
        for (int j = i + 1; j < num_items; j++) {
            if (cost[sort_indexes[i]] < cost[sort_indexes[j]]) {
                int temp = sort_indexes[i] ;
                sort_indexes[i] = sort_indexes[j] ;
                sort_indexes[j] = temp ;
            }
        }
    }

    // Calculate the maximum value that can be achieved with the given capacity
    for (int i = 0; i < num_items; i++) {
        int sorted_index = sort_indexes[i] ;

        if (max_capacity == 0) {
            free(cost) ;
            free(sort_indexes) ;
            return value ;
        }

        int capacity_intake = (weights[sorted_index] < max_capacity) ? weights[sorted_index] : max_capacity ;
        max_capacity -= capacity_intake ;
        value += capacity_intake * cost[sorted_index] ;
    }

    free(cost) ;
    free(sort_indexes) ;
    return value ;
}

int main(void) 
{
    int num_items, max_capacity ;
    scanf("%d %d", &num_items, &max_capacity) ;

    double *values = (double *) malloc(num_items * sizeof(double)) ;
    double *weights = (double *) malloc(num_items * sizeof(double)) ;

    for (int i = 0; i < num_items; i++) {
        double value, weight ;
        scanf("%lf %lf", &value, &weight) ;
        values[i] = value ;
        weights[i] = weight ;
    }

    printf("%.4lf\n", max_value(values, weights, num_items, max_capacity)) ;

    free(values) ;
    free(weights) ;
    return 0 ;
}
